llvm/lib/Frontend/OpenMP/OMP.cpp

```diff
@@ -25,6 +25,43 @@ using namespace llvm::omp;
 #define GEN_DIRECTIVES_IMPL
 #include "llvm/Frontend/OpenMP/OMP.inc"
 
+static iterator_range<ArrayRef<Directive>::iterator>
+getFirstCompositeRange(iterator_range<ArrayRef<Directive>::iterator> Leafs) {
```

Comment:
This is intended to be a general utility to find the first sequence of leaf constructs that together constitute a composite construct.  The input leafs aren't assumed to represent a legal construct.
It's also used in `isCompositeConstruct` as well, where I break up the given directive into leafs and check if the "first composite range" is equal to the sequence of leafs.

Reply:
Is there any reason for these functions to support unsupported lists of directives as input? I think it's worth having a simpler implementation over allowing all sorts of invalid lists of leaf constructs as input, unless there's a good reason why we'd need that.

Not saying that the compiler should crash in that case, but they can just return empty ranges for unsupported lists of leaf constructs rather than trying to find subsets that match what we're looking for.

Reply:
I created the concept that this function implements as a common element in implementing other functions.  This is a utility function that has a clearly defined functionality, that only depends on what is considered a "composite" construct.  It is not an API function and it doesn't have to concern itself with validating the inputs, that's left to the API functions.
There are other ways to implement the breakdown of compound constructs, yours is likely another valid approach, but I've picked this one because of its flexibility.

Reply:
I see what you mean. It still feels a bit overkill to me, but I don't have a big issue with it since it's a private function to this compilation unit, so it's not a blocker.